scripts/dev/generate_config_types.py

```diff
@@ -0,0 +1,237 @@
+    if isinstance(typ, configtypes.List):
+        return f"list[{get_type_hint(typ.valtype)}]"
+
+    if isinstance(typ, configtypes.Dict):
+        return f"Mapping[{get_type_hint(typ.keytype)}, {get_type_hint(typ.valtype)}]"
+
+    if isinstance(typ, configtypes.Bool):
+        return "bool"
+
+    valid_values = typ.get_valid_values()
+    if valid_values:
+        return f'Literal[{", ".join(literal(k) for k in valid_values.values)}]'
+    if isinstance(typ, configtypes.MappingType):
+        # note: should be redundant, but means we can get exhaustive type matching
+        return f'Literal[{", ".join(literal(k) for k in typ.MAPPING.keys())}]'
+
+    if isinstance(
+        typ,
+        (
+            configtypes.String,
+            configtypes.FontBase,
+            configtypes.QtColor,
+            configtypes.QssColor,
+            configtypes.Command,
+            configtypes.Key,
+            configtypes.SessionName,
+            configtypes.Url,
+            configtypes.File,
+            configtypes.UrlPattern,
+            configtypes.FormatString,
+            configtypes.Directory,
+            configtypes.Encoding,
+            configtypes.FuzzyUrl,
+            configtypes.SearchEngineUrl,
+            configtypes.Proxy,
+        ),
+    ):
+        return "str"
+
+    if isinstance(typ, configtypes.Int):
+        return "int"
+
+    if isinstance(typ, configtypes.Float):
+        return "float"
+
+    if isinstance(typ, configtypes.ListOrValue):
+        return f"Union[{get_type_hint(typ.valtype)}, {get_type_hint(typ.listtype)}]"
+
+    if isinstance(typ, configtypes.Perc):
+        return "Union[float, int, str]"
+
+    if isinstance(typ, configtypes.PercOrInt):
+        return "Union[int, str]"
+
+    if isinstance(typ, configtypes.Regex):
+        return "Union[str, re.Pattern[str]]"
+
+    assert_never(typ)
+
+
+def get_type_hint(typ: configtypes.BaseType) -> str:
+    """Turn a config type into a Python type annotation, e.g. `String` -> `'str'`."""
+    typ = cast(configtypes.ConfigType, typ)
+
+    inner = _get_type_hint(typ)
+
+    if typ.none_ok:
+        return f"Optional[{inner}]"
+
+    return inner
+
+
+def generate_config_types(config_data: Mapping[str, Option]) -> str:
+    """Generate the `ConfigContainer` dataclass and all nested types."""
+    output = [
+        triple_quote(
+            dedent("""
+                This file defines static types for the `c` variable in `config.py`.
+
+                This is auto-generated from the `scripts/dev/generate_config_types.py` file.
+
+                It is not intended to be used at runtime.
+
+                Example usage:
+                ```py
+                from typing import TYPE_CHECKING, cast
+
+                if TYPE_CHECKING:
+                    from qutebrowser.config.configfiles import ConfigAPI
+                    from qutebrowser.config.configcontainer import ConfigContainer
```

Comment:
In the docs, we [currently recommend](https://github.com/qutebrowser/qutebrowser/blob/main/doc/help/configuring.asciidoc#avoiding-flake8-errors):

```python
from qutebrowser.config.config import ConfigContainer
```

I wonder, can we improve things so that this still works? Would be great to make things just run without everyone who already uses types having to adjust things.

Would that maybe even give us the same autocompletion while *developing* on qutebrowser, which would be a huge plus?

Maybe in `config.py` where [`ConfigContainer` is defined](https://github.com/qutebrowser/qutebrowser/blob/v3.3.1/qutebrowser/config/config.py#L567) we could do something like:

```python
class ConfigContainer:
    ...

if TYPE_CHECKING:
    from qutebrowser.config.configcontainer import ConfigContainer
else:
    ConfigContainer = _ConfigContainer
```

and that's it?

Reply:
Yup good idea, I originally didn't want to do that just because it increases the complexity / scope of this PR but I'll look into it :)

> Would that maybe even give us the same autocompletion while developing on qutebrowser, which would be a huge plus?

Agree that would be really nice!

Reply:
I get a bunch of type errors after updating `ConfigContainer` because the actual runtime types don't match the input types as they've often been parsed / transformed in some way, e.g. for the `Url` case, the runtime type is `QUrl` but the config input is a `str`, I'm not sure what the best way to solve that would be... any ideas? generating two separate types feels like a lot of overhead...

Reply:
Hmm, right, that is unfortunate. It's because of this distinction here:

https://github.com/qutebrowser/qutebrowser/blob/ad29c973f80bd51ec5e135d3add9e8cba853c1a2/qutebrowser/config/config.py#L631-L637

where `Config.get()` gives us the value as transformed by `.to_py()` on the config type, while `Config.get_mutable_obj()` gives us the value as it's stored in the config internally...

I first thought we could do something like:

```python
CONFIGCONTAINER_INTERNAL = False
if CONFIGCONTAINER_INTERNAL:
    SelectOnRemoveType = QTabBar.SelectionBehavior
else:
    SelectOnRemoveType = Literal["prev", "next", "last-used"]
```

in `configcontainer.py` and then force `--always-true=CONFIGCONTAINER_INTERNAL` when checking qutebrowser, but it turns out that this results in mypy not knowing what the type could be when not specifying a value (it doesn't seem to catch the `CONFIGCONTAINER_INTERNAL = False`).

So I can only really think of two things, if we want to make this work for development on qutebrowser as well:

- Generate two files indeed...
- Somehow get the type system to make a distinction, similar to how [overloads](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) work maybe? But given that we're dealing with `__getattr__` here, we can't really influence anything at that point?

Maybe that's indeed something to solve outside of this PR... Though it'd still be very nice if we could make the `config.py` part at least work without everyone having to adjust their existing imports (and thus possibly not discover that this exists at all). I suppose the easiest way to do that would be to rename `ConfigContainer` to `ConfigContainerInternal` or somesuch, and then adjust all qutebrowser-internal usage to use that instead?

Reply:
> I suppose the easiest way to do that would be to rename ConfigContainer to ConfigContainerInternal or somesuch, and then adjust all qutebrowser-internal usage to use that instead?

👍 yeah that seems like the most reasonable approach to me for now